### PR TITLE
Reorganize field IO

### PR DESCRIPTION
**Pull request by RemiLehe**
This reorganizes the field IO by removing a large amount of duplication.

Basically, a lot of the code that was averaging from the nodes/edges/face to the cell centered was copied and pasted. This PR creates a dedicated function for this. Also, the corresponding function is now in a new file `FieldsIO.cpp`.

### CODE REVIEW

File: Source/Diagnostics/FieldIO.cpp [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+      MultiFab temp_dat(grids[lev],mf_avg[lev]->DistributionMap(),1,0);
+      temp_dat.setVal(0);
+
+      // MultiFab containing number of particles in each cell
+      mypc->Increment(temp_dat, lev);
+      AverageAndPackScalarField( mf_avg, temp_dat, dcomp, lev, ngrow, varnames, "part_per_cell" );
+      dcomp += 1;
+    }
+
+    if (plot_part_per_grid || plot_part_per_proc)
```
[the remark this answers is not in the document]
Comment (RemiLehe) [reply Source/Diagnostics/FieldIO.cpp]: I definitely agree. You can go on and commit the change.

**Comment by WeiqunZhang** [comment]
A few comments.

It seems to be inconsistent that the AverageAndPack* functions modify `varnames`, but not dcomp.  I think it is better to move `varnames.push_back`s out of these functions.  This will also make the function interface cleaner.  For example, `mf_avg` doesn't need to be a `Vector`.  This might also allow these functions called by `GetCellCenteredData`.

I think `amrex::average_node_to_cellcenter`, `amrex::average_face_to_cellcenter` and `amrex::average_edge_to_cellcenter` calls should take `ngrow`.  For writing plotfiles, `ngrow` is always zero.  But for `GetCellCenteredData` and `UpdateInsitu`, they call with one ghost cell.

It might be beyond the scope of this pull-request.  `UpdateInsitu` duplicates a lot of from WritePlotFile.  The main difference seems to be the number of ghost cells.  Maybe we can split `WriteAveragedFields` into two  and make both `WriteAveragedFields` and `UpdateInsitu` call the same function to do averaging of all the plotfile data.

**Comment by RemiLehe** [comment]
@WeiqunZhang These are all very good points. I will update the PR, and ping you when it is ready for review.

**Comment by WeiqunZhang** [comment]
Regarding @MaxThevenet's comment, sorry I missed it.  `NumberOfParticlesInGrid` has an MPI reduce call inside.  So it might affect performance a little big for large runs.

**Comment by RemiLehe** [comment]
@WeiqunZhang I updates the PR according to your comments. 
Would you have time to have a second pass?

File: Source/Diagnostics/FieldIO.cpp [hunk 2]
```diff
@@ -0,0 +1,299 @@
...
+        {
+            AverageAndPackScalarField( mf_avg[lev], *rho_fp[lev], dcomp, ngrow );
+            if(lev == 0) varnames.push_back("rho");
+            dcomp += 1;
+        }
+
+        if (plot_F)
+        {
+            AverageAndPackScalarField( mf_avg[lev], *F_fp[lev], dcomp, ngrow);
+            if(lev == 0) varnames.push_back("rho");
```
Comment (WeiqunZhang) [inline Source/Diagnostics/FieldIO.cpp]: There is  typo.  "rho" should be "F".

Comment (RemiLehe) [reply Source/Diagnostics/FieldIO.cpp]: Thanks for catching that!

**Pull request merged**